library/src/main/java/com/northconcepts/templatemaster/form/CrudResource.java

```diff
@@ -337,7 +338,25 @@ protected Response postEditRecordImpl(ID id, UriInfo uriInfo, ENTITY form) throw
         return gotoPath(subUrl + "/view/" + getId(form));
     }
     
+    //==========================================================================================
+    // Clone
+    //==========================================================================================
+    
+    @GET
+    @Path("/clone/{id}")
+    public Response getCloneRecord(@PathParam("id") ID id) throws Throwable {
+        if (!formDef.isAllowClone()) {
+            setErrorFlashMessage("Clone not allowed");
+            return badRequest();
+        }
 
+        ENTITY record = getClonedRecord(id);
+        if (record == null) {
+            return notFound();
```

Comment:
Let's throw an exception since this would be a bug (isAllowClone() is true, but getClonedRecord(id) returns null).  It should only be a 404 if `id` doesn't point to an existing object -- which we can test after/inside the null check.

Reply:
actually, record won't be null. if it is null, then it will throw an exception from below `TagResource` - 
`@Override
    protected Tag getClonedRecord(String id) {
        return createRecord(getRecord(id).setId(null));
    }`

should I remove this null check here?

Reply:
What makes you say it can't be null?  Won't it be null if `getClonedRecord(ID id)` isn't overridden?

Reply:
As discussed, this is not required now, so remove this null check. and resolving this conversion.